### ISSUE TITLE
NuGet support

### ISSUE THREAD

**Issue by davegaeddert** [dependencies-io/support#1]
This ought to be doable. A couple resources:

- https://docs.microsoft.com/en-us/nuget/tools/nuget-exe-cli-reference
- https://hub.docker.com/r/microsoft/dotnet/

If anybody is interested in helping, just let me know.

**Comment by Alxandr** [comment]
I'm considering taking a stab at this

**Comment by davegaeddert** [comment]
Hey @Alxandr, if you're up for it, that would be awesome!

I set up a [repo here for the collector](https://github.com/dependencies-io/collector-dotnet-nuget) -- I think that would be the best place to start. The code that is in there right now is just what we provide as an example / starting point (it is a simple python example). Forking that repo and sending pull requests would probably be the easiest place to start. In terms of getting started, these are roughly the steps that I personally would take:
- install our cli with `pip install dependencies-cli` (it's a python package that helps ease the whole process of developing)
- change the Dockerfile to "microsoft/dotnet" (I assume...)
- put any files/code that a NuGet repo would expect in [test/repo_base](https://github.com/dependencies-io/collector-dotnet-nuget/tree/master/test/repo_base) -- this will serve as an example repo for testing
- write the actual code ;)
  - mileage can vary on this, none of us are .NET users which is the reason we haven't started on this ourselves yet -- some tools make it really easy to figure parse out which packages are installed and which versions are available, others don't
  - in the end, you just need to print a JSON string in stdout that follows the format you can see here: https://github.com/dependencies-io/collector-dotnet-nuget/blob/4b213608885ab6a9fadbb09bbc7c8fc527937f9b/entrypoint.py#L19-L37
- `dependencies test -s` (available if you install dependencies-cli) will be your best friend as it will take care of starting up a new docker container and testing everything out as if it were running on dependencies.io

Let me know if you have any questions or if this is something you're interested in trying. We're just getting going on some of these things and I will tell you that you'd be the first community member to help develop a collector/actor. There will probably be some questions and I'd be happy to help as much as possible if you want to get the ball rolling. We're in the process of writing developer documentation which may help a little bit too!

**Comment by Alxandr** [comment]
I started this yesterday. Currently only the collector though. I currently have a docker container that will list all installed dependencies in a project (not in JSON format, but that's trivial). I only have left looking up the available versions, which should hopefully not be too hard.

I have a question though; If I install package `A` in my project, which has a dependency on `B`, and `B` is old, should I report `B`? Or just `A`?

**Comment by davegaeddert** [comment]
Great! Is it somewhere on GitHub where I could take a look?

_Our_ answer to your question would be to just report on `A` -- the dependencies that the user has added directly. That's how we've done everything so far and it seems to make sense.

**Comment by Alxandr** [comment]
I just pushed to here: https://github.com/YoloDev/collector-dotnet-nuget

That being said, I currently report all deps, because it's troublesome to separate out the dependences explicitly installed (root dependencies). Will have to figure something out later. Also, the code is really ugly, and PoC stage. I know how to rewrite it now though. Should only take a few hours, but not today :)

Also, I need help figuring out why the test is failing. It gives me 0 usable output :-/

**Comment by davegaeddert** [comment]
Sounds good. That's usually how we start too, and some of them still look that way.

On the tests, use `-s` (i.e. `dependencies test -s`) to show more output. Working on documenting that and a few other things...

**Comment by Alxandr** [comment]
Great. Test is now passing. I need to make it use relative paths though, as right now it reports `/repo/Test/Test.csproj` as the file.

**Comment by felixfbecker** [comment]
For Composer, we actually just added collecting transient dependencies as a feature. I think it is useful to have. Think about it - if a transient dependency (e.g. a JSON parser) receives a bugfix, that affects your direct dependency (e.g. an API client) and in turn affects your project.

Looking forward to see this!

**Comment by davegaeddert** [comment]
Yeah, I wouldn't be surprised if we continue to go the "feature" route with transitive dependencies. Especially with systems that have lockfiles of some kind, where it can be really helpful to keep that lockfile up-to-date. 

I think this was the first crack at that for Composer, for reference: https://github.com/dependencies-io/collector-php-composer/pull/12/files
Thanks to @felixfbecker!

**Comment by davegaeddert** [comment]
Yeah, I wouldn't be surprised if we continue to go the "feature" route with transitive dependencies. Especially with systems that have lockfiles of some kind, where it can be really helpful to keep that lockfile up-to-date. 

I think this was the first crack at that for Composer, for reference: https://github.com/dependencies-io/collector-php-composer/pull/12/files
Thanks to @felixfbecker!
